File: apps/api-service/internal/storage/minio_client.go
package storage

import (
	"bytes"
	"context"
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"io"
	"log/slog"
	"time"

	"beef-briefing/apps/api-service/internal/nrutil"

	"github.com/minio/minio-go/v7"
	"github.com/minio/minio-go/v7/pkg/credentials"
	"github.com/newrelic/go-agent/v3/newrelic"
)

type MinIOClient struct {
	client *minio.Client
	bucket string
	nrApp  *newrelic.Application
}

func NewMinIOClient(endpoint, accessKey, secretKey, bucket, region string, useSSL bool, nrApp *newrelic.Application) (*MinIOClient, error) {
	client, err := minio.New(endpoint, &minio.Options{
		Creds:  credentials.NewStaticV4(accessKey, secretKey, ""),
		Secure: useSSL,
		Region: region,
	})
	if err != nil {
		return nil, fmt.Errorf("failed to create MinIO client: %w", err)
	}

	mc := &MinIOClient{
		client: client,
		bucket: bucket,
		nrApp:  nrApp,
	}

	// Create bucket if it doesn't exist
	if err := mc.ensureBucket(context.Background()); err != nil {
		return nil, fmt.Errorf("failed to ensure bucket exists: %w", err)
	}

	return mc, nil
}

func (mc *MinIOClient) ensureBucket(ctx context.Context) error {
	exists, err := mc.client.BucketExists(ctx, mc.bucket)
	if err != nil {
		return fmt.Errorf("failed to check bucket existence: %w", err)
	}

	if !exists {
		slog.Info("creating MinIO bucket", "bucket", mc.bucket)
		err = mc.client.MakeBucket(ctx, mc.bucket, minio.MakeBucketOptions{})
		if err != nil {
			return fmt.Errorf("failed to create bucket: %w", err)
		}
		slog.Info("MinIO bucket created", "bucket", mc.bucket)
	}

	return nil
}

// ComputeFileHash computes the SHA256 hash of file data and returns it as a hex string.
// This can be used for content-addressable storage and deduplication.
func ComputeFileHash(data []byte) string {
	hashBytes := sha256.Sum256(data)
	return hex.EncodeToString(hashBytes[:])
}

// GenerateObjectKey generates a MinIO object key from media type and file hash.
// Format: {mediaType}/{hash[:2]}/{hash}
func GenerateObjectKey(mediaType, fileHash string) string {
	return fmt.Sprintf("%s/%s/%s", mediaType, fileHash[:2], fileHash)
}

// UploadMedia uploads media file to MinIO using content-addressable storage.
// Returns the object key and file hash. Deduplicates files with same hash.
func (mc *MinIOClient) UploadMedia(ctx context.Context, fileID string, data []byte, mimeType string, mediaType string) (string, string, error) {
	defer nrutil.StartSegment(ctx, "storage:upload-media")()

	// Compute SHA256 hash of file content
	fileHash := ComputeFileHash(data)

	// Generate object key: {mediaType}/{hash[:2]}/{hash}
	objectKey := GenerateObjectKey(mediaType, fileHash)

	// Check if object already exists (deduplication)
	_, err := mc.client.StatObject(ctx, mc.bucket, objectKey, minio.StatObjectOptions{})
	if err == nil {
		// Object already exists, skip upload
		slog.Debug("media already exists in MinIO, skipping upload",
			"object_key", objectKey,
			"file_hash", fileHash,
		)
		return objectKey, fileHash, nil
	}

	// Upload new object
	reader := bytes.NewReader(data)

	_, err = mc.client.PutObject(
		ctx,
		mc.bucket,
		objectKey,
		reader,
		int64(len(data)),
		minio.PutObjectOptions{
			ContentType: mimeType,
		},
	)
	if err != nil {
		nrutil.NoticeError(ctx, err)
		return "", "", fmt.Errorf("failed to upload to MinIO: %w", err)
	}

	slog.Debug("uploaded media to MinIO",
		"object_key", objectKey,
		"file_hash", fileHash,
		"size", len(data),
		"mime_type", mimeType,
	)

	return objectKey, fileHash, nil
}

// GetObjectURL returns the URL to access an object
func (mc *MinIOClient) GetObjectURL(objectKey string) string {
	return fmt.Sprintf("%s/%s/%s", mc.client.EndpointURL(), mc.bucket, objectKey)
}

// GetObject retrieves an object from MinIO by its key.
// Returns the object reader, content length, content type, and error.
// The caller is responsible for closing the returned reader.
func (mc *MinIOClient) GetObject(ctx context.Context, objectKey string) (io.ReadCloser, int64, string, error) {
	defer nrutil.StartSegment(ctx, "storage:get-object")()

	obj, err := mc.client.GetObject(ctx, mc.bucket, objectKey, minio.GetObjectOptions{})
	if err != nil {
		nrutil.NoticeError(ctx, err)
		return nil, 0, "", fmt.Errorf("failed to get object from MinIO: %w", err)
	}

	// Get object info for content-length and content-type
	info, err := obj.Stat()
	if err != nil {
		obj.Close()
		nrutil.NoticeError(ctx, err)
		return nil, 0, "", fmt.Errorf("failed to stat object: %w", err)
	}

	slog.Debug("retrieved object from MinIO",
		"object_key", objectKey,
		"size", info.Size,
		"content_type", info.ContentType,
	)

	return obj, info.Size, info.ContentType, nil
}

// GetPresignedURL generates a time-limited URL for accessing an object.
func (mc *MinIOClient) GetPresignedURL(ctx context.Context, objectKey string, expiry time.Duration) (string, error) {
	defer nrutil.StartSegment(ctx, "storage:get-presigned-url")()

	url, err := mc.client.PresignedGetObject(ctx, mc.bucket, objectKey, expiry, nil)
	if err != nil {
		nrutil.NoticeError(ctx, err)
		return "", fmt.Errorf("failed to generate presigned URL: %w", err)
	}

	slog.Debug("generated presigned URL",
		"object_key", objectKey,
		"expiry", expiry,
	)

	return url.String(), nil
}

// GetPresignedURLSeconds is a convenience wrapper that takes expiry in seconds.
// Implements the services.MinIOClient interface.
func (mc *MinIOClient) GetPresignedURLSeconds(ctx context.Context, objectKey string, expirySeconds int) (string, error) {
	expiry := time.Duration(expirySeconds) * time.Second
	return mc.GetPresignedURL(ctx, objectKey, expiry)
}

File: apps/card-image-generator/src/storage/minio_client.py
"""MinIO client for card image storage."""

import hashlib
import logging
from datetime import timedelta
from io import BytesIO

from minio import Minio
from minio.error import S3Error

logger = logging.getLogger(__name__)


class CardStorageClient:
    """MinIO client for uploading and retrieving card images."""

    def __init__(
        self,
        endpoint: str,
        access_key: str,
        secret_key: str,
        bucket: str,
        use_ssl: bool = False,
        region: str = "",
    ):
        self.client = Minio(
            endpoint,
            access_key=access_key,
            secret_key=secret_key,
            secure=use_ssl,
            region=region or None,
        )
        self.bucket = bucket
        self._ensure_bucket()

    def _ensure_bucket(self) -> None:
        """Create bucket if it doesn't exist."""
        try:
            if not self.client.bucket_exists(self.bucket):
                self.client.make_bucket(self.bucket)
                logger.info(f"Created bucket: {self.bucket}")
        except S3Error as e:
            logger.error(f"Failed to ensure bucket exists: {e}")
            raise

    @staticmethod
    def compute_hash(data: bytes) -> str:
        """Compute SHA256 hash of data."""
        return hashlib.sha256(data).hexdigest()

    @staticmethod
    def generate_storage_path(
        chat_id: int,
        week_start: str,
        user_id: int,
        theme: str = "gaming",
    ) -> str:
        """
        Generate storage path for a card image.

        Format: cards/{chat_id}/{week_start}/{theme}/{user_id}.png
        """
        return f"cards/{chat_id}/{week_start}/{theme}/{user_id}.png"

    def upload_card_image(
        self,
        chat_id: int,
        week_start: str,
        user_id: int,
        image_data: bytes,
        theme: str = "gaming",
    ) -> tuple[str, str, int]:
        """
        Upload card image to MinIO.

        Args:
            chat_id: Chat ID
            week_start: Week start date (YYYY-MM-DD)
            user_id: User ID
            image_data: PNG image bytes
            theme: Theme name for storage path

        Returns:
            Tuple of (storage_path, file_hash, file_size)
        """
        file_hash = self.compute_hash(image_data)
        storage_path = self.generate_storage_path(chat_id, week_start, user_id, theme)
        file_size = len(image_data)

        # Check if identical file already exists
        try:
            stat = self.client.stat_object(self.bucket, storage_path)
            existing_hash = stat.metadata.get("x-amz-meta-file-hash")
            if existing_hash == file_hash:
                logger.debug(f"Identical file exists, skipping upload: {storage_path}")
                return storage_path, file_hash, file_size
        except S3Error as e:
            # Only log unexpected errors (not "object doesn't exist")
            if e.code != "NoSuchKey":
                logger.debug(f"S3 stat check failed for {storage_path}: {e.code}")

        # Upload with metadata
        self.client.put_object(
            self.bucket,
            storage_path,
            BytesIO(image_data),
            length=file_size,
            content_type="image/png",
            metadata={"file-hash": file_hash},
        )

        logger.info(f"Uploaded card image: {storage_path} ({file_size} bytes)")
        return storage_path, file_hash, file_size

    def get_presigned_url(
        self,
        storage_path: str,
        expires_seconds: int = 3600,
    ) -> str:
        """
        Generate a presigned URL for accessing an image.

        Args:
            storage_path: Object key in bucket
            expires_seconds: URL expiration time in seconds

        Returns:
            Presigned URL string
        """
        return self.client.presigned_get_object(
            self.bucket,
            storage_path,
            expires=timedelta(seconds=expires_seconds),
        )

    def delete_card_image(self, storage_path: str) -> bool:
        """
        Delete a card image from storage.

        Returns True if deleted, False if not found.
        """
        try:
            self.client.remove_object(self.bucket, storage_path)
            logger.info(f"Deleted card image: {storage_path}")
            return True
        except S3Error as e:
            if e.code == "NoSuchKey":
                return False
            raise

    def image_exists(self, storage_path: str) -> bool:
        """Check if an image exists in storage."""
        try:
            self.client.stat_object(self.bucket, storage_path)
            return True
        except S3Error:
            return False
